src/common/transformations/src/transformations/common_optimizations/pull_through_reduce.cpp

```diff
@@ -0,0 +1,241 @@
+                                               ->get_reduction_axes()
+                                               .to_vector();
+
+        if (have_same_axes(unsqueeze_axes_val, reduce_axes_val)) {
+            return false;
+        }
+
+        if (!keep_dims) {
+            const auto unsqueeze_adjusted_axes = adjust_axes(unsqueeze_axes_val, reduce_axes_val);
+            if (unsqueeze_adjusted_axes != unsqueeze_axes_val) {
+                unsqueeze_axes_input = opset9::Constant::create(unsqueeze_axes_input->get_element_type(),
+                                                                unsqueeze_axes_input->get_shape(),
+                                                                unsqueeze_adjusted_axes);
+            }
+        }
+
+        const auto reduce_adjusted_axes = adjust_axes(reduce_axes_val, unsqueeze_axes_val);
+        if (reduce_adjusted_axes != reduce_axes_val) {
+            reduce_axes_input = opset9::Constant::create(reduce_axes_input->get_element_type(),
+                                                         reduce_axes_input->get_shape(),
+                                                         reduce_adjusted_axes);
+        }
+
+        const auto new_reduce_node = reduce_node->clone_with_new_inputs({input_node, reduce_axes_input});
+        new_reduce_node->set_friendly_name(unsqueeze_node->get_friendly_name());
+        const auto new_unsqueeze_node = unsqueeze_node->clone_with_new_inputs({new_reduce_node, unsqueeze_axes_input});
+        new_unsqueeze_node->set_friendly_name(reduce_node->get_friendly_name());
+
+        copy_runtime_info({reduce_node, unsqueeze_node}, {new_reduce_node, new_unsqueeze_node});
+        replace_node(m.get_match_root(), new_unsqueeze_node);
+
+        return true;
+    };
+
+    auto m = std::make_shared<pattern::Matcher>(reduce, matcher_name);
+    this->register_matcher(m, callback);
+}
+
+ov::pass::PullReshapeThroughReduce::PullReshapeThroughReduce() {
+    MATCHER_SCOPE(PullReshapeThroughReduce);
+
+    const auto input = pattern::any_input(pattern::has_static_shape());
+    const auto reshape_target_shape = pattern::wrap_type<opset9::Constant>();
+    const auto reshape =
+        pattern::wrap_type<opset9::Reshape>({input, reshape_target_shape}, pattern::has_static_shape());
+    const auto reduce_axes = pattern::wrap_type<opset9::Constant>();
+    const auto arithmetic_reduce = pattern::wrap_type<op::util::ArithmeticReductionKeepDims>({reshape, reduce_axes});
+    const auto logical_reduce = pattern::wrap_type<op::util::LogicalReductionKeepDims>({reshape, reduce_axes});
+    const auto reduce = std::make_shared<pattern::op::Or>(OutputVector{arithmetic_reduce, logical_reduce});
+
+    matcher_pass_callback callback = [=](pattern::Matcher& m) {
+        auto& pattern_map = m.get_pattern_value_map();
+        const auto input_node = pattern_map.at(input).get_node_shared_ptr();
```

Comment:
If `if (!reduce_axes_input)` is required move it after check, can be not used?

Reply:
moved